### PR TITLE
feat: chain snapshot

### PR DESCRIPTION
**Pull request by zhangsoledad**
Feature: 
Introduce chain snapshot, which leverage rocksdb `snapshot` and `hamt` to achieve captures point-in-time view of the chain,  get rid of `chain_state` and global lock. get 3x improve when switch fork.

* https://github.com/facebook/rocksdb/wiki/Snapshot
* https://docs.rs/im/13.0.0/im/hashmap/struct.HashMap.html

### CODE REVIEW

**Comment by quake** [comment]
hold for stress test

File: chain/src/chain.rs [hunk 1]
```diff
@@ -201,16 +233,20 @@ impl ChainService {
...
             debug!("[process_block] verification error {:?}", e);
             e.into()
         })
     }
 
-    fn insert_block(&self, block: Arc<Block>, need_verify: bool) -> Result<bool, FailureError> {
+    fn insert_block(&mut self, block: Arc<Block>, need_verify: bool) -> Result<bool, FailureError> {
+        let db_txn = self.shared.store().begin_transaction();
+        let txn_snapshot = db_txn.get_snapshot();
+        let _snapshot_tip_hash = db_txn.get_update_for_tip_hash(&txn_snapshot);
```
Comment (jjyr) [inline chain/src/chain.rs]: delete?

File: shared/src/tx_pool_ext.rs [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+                } else if self.enqueue_tx(Some(cycles), tx_size, tx) {
+                    self.update_statics_for_add_tx(tx_size, cycles);
+                }
+                Ok(cycles)
+            }),
+            Err(err) => Err(PoolError::UnresolvableTransaction(err)),
+        }
+    }
+
+    fn contains_proposed(&self, short_id: &ProposalShortId) -> bool {
```
Comment (jjyr) [inline shared/src/tx_pool_ext.rs]: The function name is confusing, maybe we can remove this function, only line 147 call this function.

File: util/proposal-table/src/lib.rs [hunk 3]
```diff
@@ -33,34 +57,22 @@ impl TxProposalTable {
...
+    pub fn finalize(
+        &mut self,
+        origin: &ProposalView,
+        number: BlockNumber,
+    ) -> (HashSet<ProposalShortId>, ProposalView) {
         let proposal_start = number.saturating_sub(self.proposal_window.farthest()) + 1;
         let proposal_end = number.saturating_sub(self.proposal_window.closest()) + 1;
 
         let mut left = self.table.split_off(&proposal_start);
         ::std::mem::swap(&mut self.table, &mut left);
```
Comment (u2) [inline util/proposal-table/src/lib.rs]:
```
        let left = self.table.split_off(&proposal_start);
        self.table = left;
```
not in this pr.

File: verification/src/block_verifier.rs [hunk 4]
```diff
@@ -8,35 +8,27 @@ use ckb_core::header::Header;
...
 
-impl<P> Verifier for BlockVerifier<P>
-where
-    P: ChainProvider + Clone,
-{
+impl<'a> Verifier for BlockVerifier<'a> {
     type Target = Block;
 
     fn verify(&self, target: &Block) -> Result<(), Error> {
-        let consensus = self.provider.consensus();
```
Comment (u2) [inline verification/src/block_verifier.rs]: :smile:

File: shared/src/lib.rs [hunk 5]
```diff
@@ -7,15 +7,15 @@
 //! - [Chain](chain::chain::Chain) represent a struct which
 //!   implement `ChainProvider`
 
-pub mod cell_set;
-pub mod chain_state;
+// pub mod cell_set;
+// pub mod chain_state;
```
[the remark this answers is not in the document]
Comment (u2) [reply shared/src/lib.rs]: can delete `cell_set` and `chain_state`.
Please port the tests in the `cell_set` to the new `cell_set`.

File: shared/src/snapshot.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+    pub fn load(&self) -> Guard<Arc<Snapshot>> {
+        self.inner.load()
+    }
+
+    pub fn store(&self, snapshot: Arc<Snapshot>) {
+        self.inner.store(snapshot);
+    }
+}
+
+pub struct Snapshot {
```
[the remark this answers is not in the document]
Comment (zhangsoledad) [reply shared/src/snapshot.rs]: I personally want to keep name `snapshot`, name confliction should not be a problem, storage snapshot should not be used directly.

File: util/types/src/core/cell.rs [hunk 7]
```diff
@@ -312,28 +309,6 @@ pub trait HeaderChecker {
     fn is_valid(&self, block_hash: &Byte32) -> bool;
 }
 
-#[derive(Default)]
```
Comment (zhangsoledad) [inline util/types/src/core/cell.rs]: only used in test, moved

**Comment by u2** [comment]
bors r=jjyr,quake,u2

**Pull request merged**